shared/utility/math/filter/eigen/ParticleFilter.h

```diff
@@ -0,0 +1,182 @@
+                    particles.col(i) = model.limit(particles.col(i));
+                }
+            }
+
+            template <typename... Args>
+            void time(const double& dt, const Args&... params) {
+                // Time update our particles
+                for (unsigned int i = 0; i < particles.cols(); ++i) {
+                    particles.col(i) = model.time(particles.col(i), dt, params...);
+                }
+
+                // Perturb them by noise
+                particles += sample_particles(StateVec::Zero(), StateVec(model.noise(dt).diagonal()), particles.cols());
+
+                // Limit the state of each particle to ensure they are still valid
+                for (unsigned int i = 0; i < particles.cols(); i++) {
+                    particles.col(i) = model.limit(particles.col(i));
+                }
+            }
+
+            template <typename MeasurementScalar, int S, int... VArgs, int... MArgs, typename... Args>
+            MeasurementScalar measure(const Eigen::Matrix<MeasurementScalar, S, 1, VArgs...>& measurement,
+                                      const Eigen::Matrix<MeasurementScalar, S, S, MArgs...>& measurement_variance,
+                                      const Args&... params) {
+
+                ParticleList candidate_particles =
+                    ParticleList::Zero(Model::size, particles.cols() + model.getRogueCount());
+                candidate_particles.leftCols(particles.cols()) = particles;
+
+                // Resample some rogues
+                for (int i = 0; i < model.getRogueCount(); ++i) {
+                    candidate_particles.col(i + particles.cols()) =
+                        particles.col(i) + model.getRogueRange().cwiseProduct(StateVec::Random() * 0.5);
+                }
+
+                Eigen::Matrix<MeasurementScalar, S, Eigen::Dynamic> differences(S, particles.cols());
+                for (int i = 0; i < candidate_particles.cols(); ++i) {
+                    differences.col(i) = model.difference(model.predict(particles.col(i), params...), measurement);
+                }
+
+                // Calculate log probabilities
+                Eigen::Matrix<MeasurementScalar, Eigen::Dynamic, 1> logits =
+                    -differences.cwiseProduct(measurement_variance.inverse() * differences).colwise().sum();
+
+                // Subtract the max log prob for numerical stability and then exponentiate
+                logits = Eigen::exp(logits.array() - logits.maxCoeff());
+
+                // Resample
+                std::discrete_distribution<> multinomial(logits.data(), logits.data() + logits.size());
+                for (unsigned int i = 0; i < particles.cols(); i++) {
+                    particles.col(i) = model.limit(candidate_particles.col(multinomial(rng)));
+                }
+                return logits.mean();
+            }
+
+            StateVec get() const {
+                // TODO take the mean, but that'll explode quaternions
```

Comment:
Any ideas on how to fix this? Maybe get should just get the most likely particle rather than the average of all particles.

Reply:
I don't know, the Armadillo one does the same thing? But is also provided a second method for only obtaining the best one. Maybe we should add that in and the user can decide what they need?

Reply:
It's then just about deciding which one should be default

Reply:
You tell me

Reply:
Added a method that should average quaternions correctly, you have to tell it where the quaternions are in the state though. If we have a way of weighting the particles (in terms of which one is better) then we could use a weighted average.

Also added a method that just returns the best particle.